### PR TITLE
Save avatar in DB

### PR DESCRIPTION
**Pull request by PKEv**
Partially addresses an issue #1656 

Question:
Whether it is necessary import old avatars?

### CODE REVIEW

**Comment by agilob** [comment]
Will this fix #2288 as well? Since avatar is saved in db, qtox will 'carry' avatars between systems in one place.

**Comment by PKEv** [comment]
yes, it will be closed

**Comment by agilob** [comment]
Ok, I will start using this for a few days and tell you the results.

**Comment by PKEv** [comment]
Thanks, I'll wait

**Comment by agilob** [comment]
So there is one issue I found so far
- [x] when loading first time, avatars from file should be read and saved in database, then removed from drive. So I lost all avatars when I started it first time:

weblate9

**Comment by PKEv** [comment]
Now avatars are common to different accounts, and if they are crossed, I can delete avatars also belong to another user...

**Comment by agilob** [comment]
Before running your branch 3/4 of the contacts had avatars.

**Comment by PKEv** [comment]
They are not lost - when you roll back to the mainline they get back, I do not delete nothing yet

**Comment by agilob** [comment]
Yup, that's the problem. When running qtox from your branch, avatars are empty (qtox is using default image). Avatars will not be set unless someone changes it in the client, which possibly means, never. Current avatars must be copied to database on start and deleted from drive after saving in DB.

**Comment by zetok** [comment]
> Current avatars must be copied to database on start and deleted from drive after saving in DB.

No, avatars shouldn't be deleted from hdd, since other clients are using that data as well.

> Avatars will not be set unless someone changes it in the client,

Err, avatars should be received every time you connect to friend - aren't they saved then?

**Comment by PKEv** [comment]
then may be import all the avatars that are in the table ALIASES  and only delete own avatar

**Comment by ProMcTagonist** [comment]
> Err, avatars should be received every time you connect to friend - aren't they saved then?

They're only received if you don't have their current avatar.

**Comment by agilob** [comment]
OK, there is only this one regression where avatars are not copied from drive to database on first start with this code. I think this should be fixed before merging so people don't lose avatars. It would go like that:

for each my contact -> check if contact has avatar in database -> check if contact has avatar on drive -> copy to database

Copying should happen only once, and that copying code will be removed after a few weeks when all users switch to the new code.

**Comment by PKEv** [comment]
Added imports at the first start.
Own avatar appear only the second start.
If you have already run this branch (the database structure has changed) for the import, you must manually remove the right column av_hash & avatar from table aliases, or the import is not fulfilled

**Comment by zetok** [comment]
c50e83c6f041d660c052241087754c14b9a5e8ee: Importing avatars doesn't really work - it imported 2 avatars, and rest was not touched.

**Comment by zetok** [comment]
5582bbbf664f55202ae088bd641197bff9b07c65: 5 imported, 2 avatars failed to import

(though that's a small sample)

**Comment by PKEv** [comment]
Need more information
try again, adding debugging code
and please send a log

**Comment by zetok** [comment]
5050fb05301a88b5f8ebcc3633d6077155c497a3: 5 imported, 2 not: https://gist.github.com/zetok/5e5032a4a80ffbfe52cc

I think that importing fails for contacts that are not in DB(?)

**Comment by agilob** [comment]
> I think that importing fails for contacts that are not in DB(?)

Which I think is ok.

**Comment by zetok** [comment]
> > I think that importing fails for contacts that are not in DB(?)
> 
> Which I think is ok.

Why would removing avatars from people that aren't in DB be ok?

**Comment by agilob** [comment]
Removing their avatars or not importing to db?

**Comment by PKEv** [comment]
Apparently these contacts do not have an avatar in dir %profile%/avatar/... 
I do not see other reason
You can send a list of files from a directory %profile%/avatar/ ?

**Comment by zetok** [comment]
@agilob 

> Removing their avatars or not importing to db?

From user perspective that's the same.

# 

@PKEv 
https://gist.github.com/zetok/ac07ba7fb09632557e7f

**Comment by PKEv** [comment]
@zetok  You really do not have an avatar for these contacts

**Comment by zetok** [comment]
> @zetok You really do not have an avatar for these contacts

Which is weird, since on qTox compiled from `master` missing avatars are shown

**Comment by PKEv** [comment]
@agilob 
i do not remove imported avatars,
it is not convenient if you have a lot of contacts with similar profiles

**Comment by zetok** [comment]
Wait - what if avatars are imported correctly, but for some reason qTox doesn't load them?

**Comment by PKEv** [comment]
@zetok I do not know what to say, they should be stored in this folder
Judging from the log, they were not imported

**Comment by PKEv** [comment]
@zetok You can run a search at a file with the name: DA6B2411E6880C6CE25DA59E4163F70C6963108DD61E2C71D725E2F59F4C7B2F
?
I wonder where he is...

**Comment by zetok** [comment]
masterPR

`master | PR`

```
zetok@galaxor ~ $ find -name 'DA6B2411E6880C6CE25DA59E4163F70C6963108DD61E2C71D725E2F59F4C7B2F*' -type f
./working/utox2/.config/tox/DA6B2411E6880C6CE25DA59E4163F70C6963108DD61E2C71D725E2F59F4C7B2F.txt
```

Huh.

Edit: from `DA6B2411E6880C6CE25DA59E4163F70C6963108DD61E2C71D725E2F59F4C7B2F.txt`: `
Hello, welcome to the Load Testing Echo Bot. This is an automatic responder that will repeat anything you say back to you. You can also use it for load testing. Say 'spam' if you want to be spammed. Say 'stop' to stop the spam. To pause the spam, set your status to away or busy.`

**Comment by PKEv** [comment]
@zetok please try again

**Comment by zetok** [comment]
f6b20e07d4555b95d33e71e6388350af3d90efb2: still missing avatars: https://gist.github.com/zetok/9ecd64931c58a46cf834

**Comment by PKEv** [comment]
:rage:
You can send a list of files from a directory %profile%?

**Comment by zetok** [comment]
https://gist.github.com/zetok/e20553f2429e12da78d5

**Comment by PKEv** [comment]
no Ideas more...

**Comment by PKEv** [comment]
Someone has no idea where else can be stored avatars?

**Comment by PKEv** [comment]
rebase and update 'delete avatar' action and add icon

**Comment by PKEv** [comment]
+ add icon to 'delete avatar' action

**Comment by agilob** [comment]
I think this code looks better and it can be merge. Could you make a final test and merge it @zetok ?

**Comment by zetok** [comment]
afb3bf35ee506a4fa56e9f3ec88e235b6eb22978: still not all avatars are shown: https://gist.github.com/zetok/c19c832f5c6ea4a0550a

**Comment by PKEv** [comment]
@zetok 
Please, tell me toxid that doesn't have avatar.
Please, try to find file with that toxid on your computer (mask: _toxid_)

**Comment by zetok** [comment]
well, there is `~/.config/tox/avatars/FB91935C998C55B6FA6C495FBD1BFFB335394CF4874FB422AFF3703B918E6A04.png`.

> Please, tell me toxid that doesn't have avatar.

That's not really doable, since a way to copy contact's PK was removed.

**Comment by ProMcTagonist** [comment]
What about groupbot?

**Comment by zetok** [comment]
@ProMcTagonist what about it?

**Comment by ProMcTagonist** [comment]
It doesn't have an avatar and its ID is public knowledge. Sorry if I misunderstood the situation.

**Comment by PKEv** [comment]
@zetok  please try again

**Comment by zetok** [comment]
@PKEv https://gist.github.com/zetok/5dcc6efc92492d5897d2 ← part of the stdout output before I've killed qTox

**Comment by PKEv** [comment]
interesting ....

**Comment by PKEv** [comment]
@zetok please try again

**Comment by zetok** [comment]
@PKEv awesome \o/
AFAICT every avatar on my test profile was successfully imported.

Note: importing seems to be very slow

I'm also going to test with my normal profile, and see how whether all avatars will be imported.

**Comment by zetok** [comment]
**very slow**: https://gist.github.com/zetok/be9e3edf2be23763691f

Good news, is that it would seem that every contact that had avatar, still has it `\o/`

File: src/persistence/historykeeper.cpp [hunk 1]
```diff
@@ -146,6 +167,21 @@ HistoryKeeper::HistoryKeeper(GenericDdInterface *db_) :
         messageID = sqlAnswer.value(0).toLongLong();
 }
 
+void HistoryKeeper::importAvatar(const QString& ownerId)
```
Comment (agilob) [inline src/persistence/historykeeper.cpp]: could you rename this method to `importAvatarToDatabase`, just `importAvatart` seems to be misleading for me.

**Comment by agilob** [comment]
This was my only comment about the branch, if all works it can be merged :)

**Comment by zetok** [comment]
> (…) if all works it can be merged :)

Well, the problem is speed - it's a serious slowdown on every start for anyone using spinning rust (hdd) who has bigger contact list.

**Comment by agilob** [comment]
but that's a one time thing, isn't it?

**Comment by zetok** [comment]
It seems to be a problem on every start.

**Comment by PKEv** [comment]
I  can make it run once.

**Comment by agilob** [comment]
Yes, it should run only once.

**Comment by PKEv** [comment]
all is ready

**Comment by agilob** [comment]
@zetok I'm ok with merging it, please test it everything works and merge before we get code conflicts.

**Pull request merged**